### PR TITLE
tdx: Adapt kata-deploy to use QEMU / OVMF from the distros

### PR DESCRIPTION
**Pull request by fidencio**
This is the second part of the work related to using the host stack from the distros that support TDX, instead of building our own.

The changes proposed here are tied to the kata-deploy script, so they can go in without any issues as long as the required CIs are passing.

Once we do changes in the runtime itself, those will require that the TDX CI passes before we get them merged.

### CODE REVIEW

**Comment by fidencio** [comment]
From a CentOS 9 Stream system:
```bash
[cloud-user@tdx-c9s kata-deploy]$ cat /etc/os-release 
NAME="CentOS Stream"
VERSION="9"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="9"
PLATFORM_ID="platform:el9"
PRETTY_NAME="CentOS Stream 9"
ANSI_COLOR="0;31"
LOGO="fedora-logo-icon"
CPE_NAME="cpe:/o:centos:centos:9"
HOME_URL="https://centos.org/"
BUG_REPORT_URL="https://bugzilla.redhat.com/"
REDHAT_SUPPORT_PRODUCT="Red Hat Enterprise Linux 9"
REDHAT_SUPPORT_PRODUCT_VERSION="CentOS Stream"


[cloud-user@tdx-c9s kata-deploy]$ sudo k3s kubectl -n kube-system logs kata-deploy-lcxf9
Environment variables passed to this script
* NODE_NAME: tdx-c9s
* DEBUG: false
* SHIMS: clh cloud-hypervisor dragonball fc qemu qemu-coco-dev qemu-nvidia-gpu qemu-sev qemu-snp qemu-tdx stratovirt
* DEFAULT_SHIM: qemu
* CREATE_RUNTIMECLASSES: false
* CREATE_DEFAULT_RUNTIMECLASS: false
* ALLOWED_HYPERVISOR_ANNOTATIONS: 
* SNAPSHOTTER_HANDLER_MAPPING: 
* AGENT_HTTPS_PROXY: 
* AGENT_NO_PROXY: 
time="2024-05-08T15:59:04Z" level=info msg="Acquiring lock file /var/lib/rancher/k3s/data/.lock"
time="2024-05-08T15:59:04Z" level=info msg="Preparing data dir /var/lib/rancher/k3s/data/b159f6e26663d8c92285e7bc4a6881d85bd8c81efc55eb2cf191c54100387fbb"
Validating the snapshotter-handler mapping: ""
No snapshotter has been requested, using the default value from containerd
copying kata artifacts onto host
INFO: In order to use the tdx related runtime classes, ensure TDX is properly configured for centos 9 by following the instructions provided at: https://sigs.centos.org/virt/tdx
Creating the default shim-v2 binary
Add Kata Containers as a supported runtime for containerd


[cloud-user@tdx-c9s kata-deploy]$ for i in `ls /opt/kata/share/defaults/kata-containers/*.toml`; do echo $i && cat $i | grep -E "^path =|^firmware =" | grep -v "/opt/kata/" | grep -v "^firmware = \"\"" && echo ""; done 
/opt/kata/share/defaults/kata-containers/configuration-acrn.toml
/opt/kata/share/defaults/kata-containers/configuration-clh.toml
/opt/kata/share/defaults/kata-containers/configuration-fc.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-coco-dev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-tdx.toml
path = "PLACEHOLDER_FOR_DISTRO_QEMU_WITH_TDX_SUPPORT"
firmware = "PLACEHOLDER_FOR_DISTRO_OVMF_WITH_TDX_SUPPORT"

/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-se.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-sev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-tdx.toml
path = "/usr/libexec/qemu-kvm"
firmware = "/usr/share/edk2/ovmf/OVMF.inteltdx.fd"

/opt/kata/share/defaults/kata-containers/configuration-qemu.toml
/opt/kata/share/defaults/kata-containers/configuration-remote.toml
/opt/kata/share/defaults/kata-containers/configuration-stratovirt.toml
/opt/kata/share/defaults/kata-containers/configuration.toml
```

The only reason you see:
```bash
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-tdx.toml
path = "PLACEHOLDER_FOR_DISTRO_QEMU_WITH_TDX_SUPPORT"
firmware = "PLACEHOLDER_FOR_DISTRO_OVMF_WITH_TDX_SUPPORT"
```

Is because https://github.com/kata-containers/kata-containers/pull/9484 is not yet ready to be merged.

**Comment by fidencio** [comment]
From a Ubuntu 24.04 system:
```bash
ubuntu@tdx-ubuntu:~$ cat /etc/os-release 
PRETTY_NAME="Ubuntu 24.04 LTS"
NAME="Ubuntu"
VERSION_ID="24.04"
VERSION="24.04 LTS (Noble Numbat)"
VERSION_CODENAME=noble
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=noble
LOGO=ubuntu-logo



ubuntu@tdx-ubuntu:~$ sudo k3s kubectl -n kube-system logs kata-deploy-zzjr5
Environment variables passed to this script
* NODE_NAME: tdx-ubuntu
* DEBUG: false
* SHIMS: clh cloud-hypervisor dragonball fc qemu qemu-coco-dev qemu-nvidia-gpu qemu-sev qemu-snp qemu-tdx stratovirt
* DEFAULT_SHIM: qemu
* CREATE_RUNTIMECLASSES: false
* CREATE_DEFAULT_RUNTIMECLASS: false
* ALLOWED_HYPERVISOR_ANNOTATIONS: 
* SNAPSHOTTER_HANDLER_MAPPING: 
* AGENT_HTTPS_PROXY: 
* AGENT_NO_PROXY: 
time="2024-05-08T16:11:06Z" level=info msg="Acquiring lock file /var/lib/rancher/k3s/data/.lock"
time="2024-05-08T16:11:06Z" level=info msg="Preparing data dir /var/lib/rancher/k3s/data/b159f6e26663d8c92285e7bc4a6881d85bd8c81efc55eb2cf191c54100387fbb"
Validating the snapshotter-handler mapping: ""
No snapshotter has been requested, using the default value from containerd
copying kata artifacts onto host
INFO: In order to use the tdx related runtime classes, ensure TDX is properly configured for ubuntu 24.04 by following the instructions provided at: https://github.com/canonical/tdx/tree/noble-24.04
Creating the default shim-v2 binary
Add Kata Containers as a supported runtime for containerd
node/tdx-ubuntu not labeled



ubuntu@tdx-ubuntu:~$ for i in `ls /opt/kata/share/defaults/kata-containers/*.toml`; do echo $i && cat $i | grep -E "^path =|^firmware =" | grep -v "/opt/kata/" | grep -v "^firmware = \"\"" && echo ""; done 
/opt/kata/share/defaults/kata-containers/configuration-acrn.toml
/opt/kata/share/defaults/kata-containers/configuration-clh.toml
/opt/kata/share/defaults/kata-containers/configuration-fc.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-coco-dev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-tdx.toml
path = "PLACEHOLDER_FOR_DISTRO_QEMU_WITH_TDX_SUPPORT"
firmware = "PLACEHOLDER_FOR_DISTRO_OVMF_WITH_TDX_SUPPORT"

/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-se.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-sev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-tdx.toml
path = "/usr/bin/qemu-system-x86_64"
firmware = "/usr/share/ovmf/OVMF.fd"

/opt/kata/share/defaults/kata-containers/configuration-qemu.toml
/opt/kata/share/defaults/kata-containers/configuration-remote.toml
/opt/kata/share/defaults/kata-containers/configuration-stratovirt.toml
/opt/kata/share/defaults/kata-containers/configuration.toml
```

The same comment about the NVIDIA runtime class applies here.

**Comment by fidencio** [comment]
And finally on a Ubuntu 22.04 system.
```bash
ubuntu@k3s:~/kata-containers$ cat /etc/os-release 
PRETTY_NAME="Ubuntu 22.04.2 LTS"
NAME="Ubuntu"
VERSION_ID="22.04"
VERSION="22.04.2 LTS (Jammy Jellyfish)"
VERSION_CODENAME=jammy
ID=ubuntu
ID_LIKE=debian
HOME_URL="https://www.ubuntu.com/"
SUPPORT_URL="https://help.ubuntu.com/"
BUG_REPORT_URL="https://bugs.launchpad.net/ubuntu/"
PRIVACY_POLICY_URL="https://www.ubuntu.com/legal/terms-and-policies/privacy-policy"
UBUNTU_CODENAME=jammy


ubuntu@k3s:~/kata-containers$ kubectl -n kube-system logs kata-deploy-mthfs
Environment variables passed to this script
* NODE_NAME: k3s
* DEBUG: false
* SHIMS: clh cloud-hypervisor dragonball fc qemu qemu-coco-dev qemu-nvidia-gpu qemu-sev qemu-snp qemu-tdx stratovirt
* DEFAULT_SHIM: qemu
* CREATE_RUNTIMECLASSES: false
* CREATE_DEFAULT_RUNTIMECLASS: false
* ALLOWED_HYPERVISOR_ANNOTATIONS: 
* SNAPSHOTTER_HANDLER_MAPPING: 
* AGENT_HTTPS_PROXY: 
* AGENT_NO_PROXY: 
Validating the snapshotter-handler mapping: ""
No snapshotter has been requested, using the default value from containerd
copying kata artifacts onto host
WARN: Distro ubuntu 22.04 does not support TDX and the TDX related runtime classes will not work in your cluster!
Creating the default shim-v2 binary
Add Kata Containers as a supported runtime for containerd
node/k3s labeled


ubuntu@k3s:~/kata-containers$ for i in `ls /opt/kata/share/defaults/kata-containers/*.toml`; do echo $i && cat $i | grep -E "^path =|^firmware =" | grep -v "/opt/kata/" | grep -v "^firmware = \"\"" && echo ""; done
/opt/kata/share/defaults/kata-containers/configuration-acrn.toml
/opt/kata/share/defaults/kata-containers/configuration-clh.toml
/opt/kata/share/defaults/kata-containers/configuration-fc.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-coco-dev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu-tdx.toml
path = "PLACEHOLDER_FOR_DISTRO_QEMU_WITH_TDX_SUPPORT"
firmware = "PLACEHOLDER_FOR_DISTRO_OVMF_WITH_TDX_SUPPORT"

/opt/kata/share/defaults/kata-containers/configuration-qemu-nvidia-gpu.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-se.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-sev.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-snp.toml
/opt/kata/share/defaults/kata-containers/configuration-qemu-tdx.toml
path = "PLACEHOLDER_FOR_DISTRO_QEMU_WITH_TDX_SUPPORT"
firmware = "PLACEHOLDER_FOR_DISTRO_OVMF_WITH_TDX_SUPPORT"

/opt/kata/share/defaults/kata-containers/configuration-qemu.toml
/opt/kata/share/defaults/kata-containers/configuration-remote.toml
/opt/kata/share/defaults/kata-containers/configuration-stratovirt.toml
/opt/kata/share/defaults/kata-containers/configuration.toml

```

**Comment by fidencio** [comment]
Having or not the config file and / or  runtime-class deployed is out of the scope of this PR, as we currently deploy, by default, config files that may or may not work (acrn, firecracker) and runtime classes (such as firecracker).

**Comment by fidencio** [comment]
As I was able to bring back the TDX CI, I've added here the patches that should make QEMU work.
I'm also adding here the patch for skipping the pull-image tests on TEEs, just to let the CI run. :-)

**Review by GabyCT — APPROVED** [review]

**Review by danmihai1 — APPROVED** [review]

File: src/runtime/Makefile [hunk 1]
```diff
@@ -175,7 +176,8 @@ HYPERVISORS := $(HYPERVISOR_ACRN) $(HYPERVISOR_FC) $(HYPERVISOR_QEMU) $(HYPERVIS
 QEMUPATH := $(QEMUBINDIR)/$(QEMUCMD)
 QEMUVALIDHYPERVISORPATHS := [\"$(QEMUPATH)\"]
 
-QEMUTDXPATH := $(QEMUBINDIR)/$(QEMUTDXCMD)
+#QEMUTDXPATH := $(QEMUBINDIR)/$(QEMUTDXCMD)
```
Comment (stevenhorsman) [inline src/runtime/Makefile:179]: Are you leaving the originals here to make it easy to revert once we have upstream qemu supported? 

PS: I don't have an issue with this, just checking.

File: tools/packaging/kata-deploy/scripts/kata-deploy.sh [hunk 2]
```diff
@@ -39,6 +39,16 @@ die() {
         exit 1
 }
 
+warn() {
+        msg="$*"
+        echo "WARN: $msg" >&2
```
Comment (stevenhorsman) [inline tools/packaging/kata-deploy/scripts/kata-deploy.sh:44]: nit: 8 spaces rather than a tab, which is inconsistent with the rest of the file

**Review by stevenhorsman — APPROVED** [review]
Generally looks good and better than I had imagined it, just a couple of minor questions/comments

File: src/runtime/Makefile (hunk 1 above)
Comment (fidencio) [reply src/runtime/Makefile:179]: That's exactly the idea. :-)

File: tools/packaging/kata-deploy/scripts/kata-deploy.sh (hunk 2 above)
Comment (fidencio) [reply tools/packaging/kata-deploy/scripts/kata-deploy.sh:44]: I promise I will fix it in a follow-up, next time I touch this file.
For now, let's save the planet and avoid re-running the CIs. :-)

**Pull request merged**